app/gui2/src/util/config.ts

```diff
@@ -0,0 +1,477 @@
+        this.addOption(name, option)
+      }
+    }
+    const groups = cfg?.groups
+    if (groups != null) {
+      for (const [name, group] of Object.entries(groups)) {
+        this.addGroup(name, group)
+      }
+    }
+  }
+
+  addOption(name: string, option: AnyOption) {
+    const existingOption = this.options[name]
+    if (existingOption != null) {
+      console.error(`Duplicate config option found '${existingOption.qualifiedName()}'.`)
+    }
+    const options = this.options as OptionsRecord
+    options[name] = option
+    option.name = name
+  }
+
+  addGroup(name: string, group: AnyGroup) {
+    const existingGroup = this.groups[name]
+    if (existingGroup != null) {
+      existingGroup.merge(group)
+    } else {
+      const groups = this.groups as GroupsRecord
+      groups[name] = group
+      group.name = name
+    }
+    group.setPath([name])
+  }
+
+  /** Set the path of this group. If this group was placed in another group, the path will contain
+   * all the parent group names. */
+  setPath(path: string[]) {
+    for (const option of Object.values(this.options)) {
+      option.path = path
+    }
+    for (const [name, group] of Object.entries(this.groups)) {
+      group.setPath(path.concat([name]))
+    }
+  }
+
+  /** Return a deep copy of this group definition. */
+  clone(): this {
+    const result: AnyGroup = new Group()
+    for (const [name, group] of Object.entries(this.groups)) {
+      result.groups[name] = group.clone()
+    }
+    for (const [name, option] of Object.entries(this.options)) {
+      result.options[name] = option.clone()
+    }
+    return result as this
+  }
+
+  /** Merge this group definition with another group definition. Returns a deeply merged group. In
+   * case the argument will override some options, errors will be logged. */
+  merge<Other extends AnyGroup>(other?: Other | null): this & Other {
+    if (other == null) {
+      return this as this & Other
+    } else {
+      const result: AnyGroup = new Group()
+      Object.assign(result.groups, this.groups)
+      for (const [otherGroupName, otherGroup] of Object.entries(other.groups)) {
+        const group = result.groups[otherGroupName]
+        if (group == null) {
+          result.groups[otherGroupName] = otherGroup
+        } else {
+          result.groups[otherGroupName] = group.merge(otherGroup)
+        }
+      }
+      Object.assign(result.options, this.options)
+      for (const [otherOptionName, otherOption] of Object.entries(other.options)) {
+        const option = result.options[otherOptionName]
+        if (option != null) {
+          console.error(`Duplicate config option found '${option.qualifiedName()}'.`)
+        }
+        result.options[otherOptionName] = otherOption
+      }
+      result.name = this.name
+      result.description = this.description
+      return result as this & Other
+    }
+  }
+
+  load(config: StringConfig, stack: string[] = []): string[] {
```

Comment:
This is not loading anything, it is just input validation. Though it probably SHOULD be loading - a static method accepting unknown input and returning the `Result` of either loaded group or a list of unrecognized names.

Reply:
technically it *does* load via `option.load(value)` below... it *is* really janky though